R/merge_sce_list.R

```diff
@@ -0,0 +1,138 @@
+#' @param retain_coldata_columns_main A vector of colData columns which should be retained
+#'  in the main experiment of the final merged SCE object.
+#' @param retain_coldata_columns_altexps A vector of colData columns which should be retained
+#'  in the altExps experiments of the final merged SCE object.
+#' @param retain_altexps A logical indicating whether any present altExp slots should
+#'   be retained in the merged SCE. Default is `TRUE`.
+#' @param preserve_rowdata_columns A vector of column names that appear in originating
+#'  SingleCellExperiment objects' rowData slots which should not be renamed with
+#'  the given SingleCellExperiment object name. These are generally columns which are
+#'  not specific to the given library's preparation or statistics. For example,
+#'  such an array might contain items like "Gene", "ensembl_ids", "gene_symbol", etc.
+#'
+#' @return A SingleCellExperiment object containing all SingleCellExperiment objects
+#'   present in the inputted list
+#' @export
+#'
+#' @import SingleCellExperiment
+merge_sce_list <- function(sce_list = list(),
+                           batch_column = "batch",
+                           retain_altexps = TRUE,
+                           retain_coldata_columns_main = c("sum",
+                                                           "detected",
+                                                           "total",
+                                                           "subsets_mito_sum",
+                                                           "subsets_mito_detected",
+                                                           "subsets_mito_percent",
+                                                           "miQC_pass"),
+                           retain_coldata_columns_altexps = c(),
+                           preserve_rowdata_columns = NULL) {
+  # Ensure `sce_list` is named  ----------------------
+  if (is.null(names(sce_list))) {
+    stop("Individual SingleCellExperiment objects in `sce_list` must be named.")
+  }
+  # Ensure `sce_list` has >=2 items --------------------------------------------
+  if (length(sce_list) < 2) {
+    stop("The `sce_list` must contain 2 or more SingleCellExperiment objects to merge.")
+  }
+
+  # Subset SCEs to shared features and ensure appropriate naming ------------------
+
+  # First, obtain intersection among all SCE objects
+  shared_features <- sce_list |>
+    purrr::map(rownames) |>
+    purrr::reduce(intersect)
+
+
+  process_columns <- function(sce, sce_name, retain_columns) {
+    # helper function to rename and subset columns in an SCE object when
+    # formatting for merging
+
+    ##### rowData #####
+    # Add `sce_name` ID to rowData column names except for those
+    #  present in `preserve_rowdata_columns`
+    original_colnames <- colnames(rowData(sce))
+
+    colnames(rowData(sce)) <- ifelse(
+      original_colnames %in% preserve_rowdata_columns,
```

Comment:
`preserve_rowdata_columns` is not an argument to the function, but it should be. (I know that this works because it is an internal function,  but I am hugely not a fan of functions going outside their defined scope, no matter what. It makes for unexpected results too often.)